tensorboard/webapp/routes/core_deeplink_provider_test.ts

```diff
@@ -0,0 +1,227 @@
+    provider
+      .serializeStateToQueryParams(store)
+      .pipe(
+        // Skip the initial bootstrap.
+        skip(1)
+      )
+      .subscribe((queryParams) => {
+        queryParamsSerialized.push(queryParams);
+      });
+  });
+
+  describe('time series', () => {
+    it('serializes pinned card state when store updates', () => {
+      store.overrideSelector(selectors.getPinnedCardsWithMetadata, [
+        {
+          cardId: 'card1',
+          plugin: PluginType.SCALARS,
+          tag: 'accuracy',
+          runId: null,
+        },
+      ]);
+      store.overrideSelector(selectors.getUnresolvedImportedPinnedCards, [
+        {
+          plugin: PluginType.SCALARS,
+          tag: 'loss',
+        },
+      ]);
+      store.refreshState();
+
+      expect(queryParamsSerialized[queryParamsSerialized.length - 1]).toEqual([
+        {
+          key: 'pinnedCards',
+          value:
+            '[{"plugin":"scalars","tag":"accuracy"},{"plugin":"scalars","tag":"loss"}]',
+        },
+      ]);
+
+      store.overrideSelector(selectors.getPinnedCardsWithMetadata, [
+        {
+          cardId: 'card1',
+          plugin: PluginType.SCALARS,
+          tag: 'accuracy2',
+          runId: null,
+        },
+      ]);
+      store.overrideSelector(selectors.getUnresolvedImportedPinnedCards, [
+        {
+          plugin: PluginType.SCALARS,
+          tag: 'loss2',
+        },
+      ]);
+      store.refreshState();
+
+      expect(queryParamsSerialized[queryParamsSerialized.length - 1]).toEqual([
+        {
+          key: 'pinnedCards',
+          value:
+            '[{"plugin":"scalars","tag":"accuracy2"},{"plugin":"scalars","tag":"loss2"}]',
+        },
+      ]);
+    });
+
+    it('serializes nothing when states are empty', () => {
+      store.overrideSelector(selectors.getPinnedCardsWithMetadata, []);
+      store.overrideSelector(selectors.getUnresolvedImportedPinnedCards, []);
+      store.refreshState();
+
+      expect(queryParamsSerialized[queryParamsSerialized.length - 1]).toEqual(
+        []
+      );
+    });
+
+    it('deserializes empty pinned cards', () => {
+      const state = provider.deserializeQueryParams([]);
+
+      expect(state).toEqual({metrics: {pinnedCards: []}});
+    });
+
+    it('deserializes valid pinned cards', () => {
+      const state = provider.deserializeQueryParams([
+        {
+          key: 'pinnedCards',
+          value:
+            '[{"plugin":"scalars","tag":"accuracy"},{"plugin":"images","tag":"loss","runId":"exp1/123","sample":5}]',
+        },
+      ]);
+
+      expect(state).toEqual({
+        metrics: {
+          pinnedCards: [
+            {plugin: PluginType.SCALARS, tag: 'accuracy'},
+            {
+              plugin: PluginType.IMAGES,
+              tag: 'loss',
+              runId: 'exp1/123',
+              sample: 5,
+            },
+          ],
+        },
+      });
+    });
+
+    it('sanitizes pinned cards on deserialization', () => {
```

Comment:
My preference would be that each of these is there own test case for several reasons: It's best practice to avoid control structures in tests; A particular test failure would be easier to identify by line number; tests would be easier to debug (wouldn't have to wait for any particular test cases's turn in the for loop).

It wouldn't be very much more verbose, if at all.

Reply:
But I see now this exact use case ("data-driven testing") is the one exception to the advice not to avoid control structures in testing:

go/unit-test-practices?polyglot=java#control-structures

Reply:
I think the current pattern fits "data-driven testing", though I can see why we'd want to split them up if the test had even a bit more complexity.  Will keep as is, but I'll keep this in mind in the future.